### PR TITLE
fix(file-upload): add mime types to file chooser to enable uploads on iOS

### PR DESCRIPTION
**Pull request by zyv**
I've added mime types as you suggested, and it indeed seems to fix the problem on iOS. I didn't test in any other browsers.

The `application/octet-stream` might be too unspecific, but I'm not sure what else I could do to accept binary formats like FF or Garmin :(

### CODE REVIEW

**Comment by rdamazio** [comment]
> The application/octet-stream might be too unspecific, but I'm not sure what else I could do to accept binary formats like FF or Garmin :(

afaik having `application/octet-stream` is pretty much equivalent to removing the accept field altogether.

The apps will register the mime types they support (with associated extensions) -
https://developer.apple.com/documentation/uniformtypeidentifiers/defining-file-and-data-types-for-your-app

So you need to find out what those are on iOS, which I think you can do by looking in the IPA file:
https://apple.stackexchange.com/questions/415575/how-to-download-an-app-file-from-the-app-store

**Comment by zyv** [comment]
> > The application/octet-stream might be too unspecific, but I'm not sure what else I could do to accept binary formats like FF or Garmin :(
> 
> afaik having `application/octet-stream` is pretty much equivalent to removing the accept field altogether.

It most certainly isn't, but it absolutely does allow unwanted plain binary files, which is why I'm bringing it up. I don't think I can convey a feeling of what it really looks like. I think it's best for you to actually try it on your machine to get a feeling of how good or bad it is.

> The apps will register the mime types they support (with associated extensions) - https://developer.apple.com/documentation/uniformtypeidentifiers/defining-file-and-data-types-for-your-app
> 
> So you need to find out what those are on iOS, which I think you can do by looking in the IPA file: https://apple.stackexchange.com/questions/415575/how-to-download-an-app-file-from-the-app-store

You bet I know how to do this, but how does this help? A formal public type registration is missing:

https://www.iana.org/assignments/media-types/media-types.xhtml

Anyway, for the record, they most certainly do not register a type-based handler, only an extension-based one (`public.mime-type` tag is missing):

```xml
<dict>                       
  <key>UTTypeIdentifier</key>
  <string>com.foreflight.fmd</string>
  <key>UTTypeDescription</key>
  <string>FMD</string>
  <key>UTTypeConformsTo</key>
  <array>
    <string>public.data</string>
  </array>
  <key>UTTypeTagSpecification</key>
  <dict>
    <key>public.filename-extension</key>
    <array>
      <string>fmd</string>
    </array>
  </dict>
</dict>
```

Also, how about ACE and other formats like Garmin Pilot?

What I was thinking about was maybe only adding types on Safari, but I didn't want to voice this suggestion.

**Comment by rdamazio** [comment]
> they most certainly do not register a type-based handler
> What I was thinking about was maybe only adding types on Safari, but I didn't want to voice this suggestion.

If there's no way to properly filter for the extensions, how about keeping it simpler, and dropping the accept altogether (allowing any file), but only on iOS/iPadOS Safari? The import will still fail if they pick an invalid file.

(there'a already a browser detection thing used in welcome.component.ts, you could use that for this too)

Why I think that's ok: people usually don't have a ton of loose files on their iPad, so the filtering doesn't make that much difference.

It's also been reported here: https://github.com/mdn/browser-compat-data/issues/26043 and on http://www.openradar.me/36726477

**Comment by zyv** [comment]
> If there's no way to properly filter for the extensions, how about keeping it simpler, and dropping the accept altogether (allowing any file), but only on iOS/iPadOS Safari? The import will still fail if they pick an invalid file. 
> 
> Why I think that's ok: people usually don't have a ton of loose files on their iPad, so the filtering doesn't make that much difference.

We could do that, but desktop Safari is also affected, although it behaves in a different and even more confusing manner. After some clicking around, files become selectable. So we could drop it for Safari in general, but the desktop experience will be affected in quite an annoying way. What do you think about limiting it to Safari on iOS and desktop?

**Comment by rdamazio** [comment]
> What do you think about limiting it to Safari on iOS and desktop?

I'm ok with that if you are.

**Comment by zyv** [comment]
> > What do you think about limiting it to Safari on iOS and desktop?
> 
> I'm ok with that if you are.

Alright, this is then what I'd go for. I will appreciate an approval. Thanks!

**Comment by rdamazio** [comment]
>  I will appreciate an approval. 

I don't see an updated version of the commits?

**Comment by rdamazio** [comment]
Or maybe I misunderstood what you mean by "limiting it to Safari on iOS and desktop" - I understood you meant removing accept altogether only on those, is that what you meant?

**Comment by zyv** [comment]
> Or maybe I misunderstood what you mean by "limiting it to Safari on iOS and desktop" - I understood you meant removing accept altogether only on those, is that what you meant?

Oh no, I see now that I've completely misunderstood what you meant. I meant I'd **add** accept only on Safari. I don't like the idea of removing the accept completely, because it would broaden the selection such that I can select any files at all, and I find this annoying on desktop. If the MIME types are added, even rather unrestrictive ones, it does look much better and it's much harder to misclick. This is why I added the second commit.

File: src/model/formats/garmin-pilot-format.ts [hunk 1]
```diff
@@ -4,6 +4,8 @@ import { GarminPilotReader } from './garmin-pilot-reader';
 import { GarminPilotWriter } from './garmin-pilot-writer';
 
 export class GarminPilotFormat extends AbstractChecklistFormat {
+  public readonly mimeType = 'application/gzip';
```
Comment (rdamazio) [inline src/model/formats/garmin-pilot-format.ts]: Does this mime type actually let you select .gplt files?

**Review by rdamazio — DISMISSED** [review]
Ah, ok - yes, I'm fine with that too. Just a minor comment.

File: src/model/formats/garmin-pilot-format.ts (hunk 1 above)
Comment (zyv) [reply src/model/formats/garmin-pilot-format.ts]: You know what, I did some more careful testing (I had to empty caches explicitly every time I make a change to completely reload the app), and here are the results:

On macOS with the latest Safari, the selector works properly. It used to behave as I described, but I think that if with the latest updates it's working, it’s better not to change anything.

On iOS, the situation is interesting indeed. If I leave out accept, or accept `*` or something like this, then, whenever the box is clicked, an annoying submenu is shown which either asks to make a photo, select a picture from the camera roll, or upload a file.

However, if one MIME type is added to the extensions (no matter which one), a file chooser opens directly. The files are, unfortunately, unfiltered, but as you said, it's probably unlikely that you have many files anyway and it's better to be able to select them than have to hack around with drag and drop.

So my latest approach is closer to what you suggested, that is to leave everything as is, except for adding `application/octet-stream` only on iOS. Would you agree to that?

File: src/app/checklists/file-upload/file-upload.component.ts [hunk 2]
```diff
@@ -13,7 +14,10 @@ import { FORMAT_REGISTRY, parseChecklistFile } from '../../../model/formats/form
 })
 export class ChecklistFileUploadComponent {
   readonly fileUploaded = output<ChecklistFile>();
-  protected readonly _acceptExtensions = FORMAT_REGISTRY.getSupportedInputExtensions();
+  protected readonly _deviceService = inject(DeviceDetectorService);
+  protected readonly _acceptExtensions =
+    FORMAT_REGISTRY.getSupportedInputExtensions() +
+    (this._deviceService.deviceInfo().os === OS.IOS ? ', application/octet-stream' : '');
```
Comment (rdamazio) [inline src/app/checklists/file-upload/file-upload.component.ts:23]: Add a comment explaining why.

Comment (rdamazio) [inline src/app/checklists/file-upload/file-upload.component.ts:23]: Doesn't this also match Chrome and other browsers on IOS? Is the situation the same there?

File: src/model/formats/garmin-pilot-format.ts (hunk 1 above)
Comment (rdamazio) [reply src/model/formats/garmin-pilot-format.ts]: > So my latest approach is closer to what you suggested, that is to leave everything as is, except for adding `application/octet-stream` only on iOS. Would you agree to that?

Yup

File: src/app/checklists/file-upload/file-upload.component.ts (hunk 2 above)
Comment (zyv) [reply src/app/checklists/file-upload/file-upload.component.ts:23]: Done

Comment (zyv) [reply src/app/checklists/file-upload/file-upload.component.ts:23]: That's correct, but all browsers on iOS must use Apple’s engine by policy, so the behavior should be identical. I've tested with Orion and Chrome, and indeed they behave exactly the same. So I think this is the right thing to do.

**Review by rdamazio — APPROVED** [review]
